File: swift-sdks/Sources/SuiSwiftSDK/Transactions/Executor/TransactionExecutor.swift
import Foundation

public enum TransactionExecutionError: Error, LocalizedError {
    case queueFull
    case queueClosed
    case transactionFailed(String)
    case executorBusy

    public var errorDescription: String? {
        switch self {
        case .queueFull: return "Transaction queue is full"
        case .queueClosed: return "Transaction queue is closed"
        case .transactionFailed(let msg): return "Transaction failed: \(msg)"
        case .executorBusy: return "Transaction executor is busy"
        }
    }
}

public struct TransactionResult {
    public let success: Bool
    public let error: String?

    public init(success: Bool, error: String? = nil) {
        self.success = success
        self.error = error
    }

    public static func success() -> TransactionResult {
        return TransactionResult(success: true)
    }

    public static func failure(_ error: String) -> TransactionResult {
        return TransactionResult(success: false, error: error)
    }
}

public class ObjectCache {
    private var cache: [String: [String: Any]] = [:]
    private var ownedObjects: [String: [String: Any]] = [:]
    private var customCache: [String: [String: Any]] = [:]

    public init() {}

    public func set(_ id: String, object: [String: Any]) {
        cache[id] = object
    }

    public func get(_ id: String) -> [String: Any]? {
        return cache[id]
    }

    public func delete(_ id: String) {
        cache.removeValue(forKey: id)
        ownedObjects.removeValue(forKey: id)
        customCache.removeValue(forKey: id)
    }

    public func clearOwnedObjects() {
        ownedObjects.removeAll()
    }

    public func clearCustom() {
        customCache.removeAll()
    }

    public func getOwnedObject(_ id: String) -> [String: Any]? {
        return ownedObjects[id]
    }

    public func setOwnedObject(_ id: String, object: [String: Any]) {
        ownedObjects[id] = object
    }

    public func getCustom(_ key: String) -> [String: Any]? {
        return customCache[key]
    }

    public func setCustom(_ key: String, value: [String: Any]) {
        customCache[key] = value
    }

    public func deleteCustom(_ key: String) {
        customCache.removeValue(forKey: key)
    }
}

public class CachingTransactionExecutor {
    private let cache: ObjectCache

    public init(cache: ObjectCache = ObjectCache()) {
        self.cache = cache
    }

    public func execute(_ transaction: SuiTransaction) async throws -> TransactionResult {
        return TransactionResult.success()
    }

    public func applyEffects(_ effects: [String: Any]) {
        if let changedObjects = effects["changedObjects"] as? [[String: Any]] {
            for changedObj in changedObjects {
                if let objectId = changedObj["objectId"] as? String,
                   let outputState = changedObj["outputState"] as? [String: Any],
                   outputState["ObjectWrite"] != nil {
                    cache.setOwnedObject(objectId, object: changedObj)
                }
            }
        }
    }

    public func reset() {
        cache.clearOwnedObjects()
        cache.clearCustom()
    }

    public func getCache() -> ObjectCache {
        return cache
    }
}

public class SerialQueue {
    private var taskQueue: [() async throws -> Void] = []
    private var isProcessing = false

    public init() {}

    public func runTask<T>(_ task: @escaping () async throws -> T) async throws -> T {
        return try await task()
    }
}

public class ParallelQueue {
    private let maxConcurrency: Int
    private var activeTasks = 0
    private var taskQueue: [() async throws -> Void] = []
    private var continuations: [CheckedContinuation<Void, Never>] = []

    public init(maxConcurrency: Int = 4) {
        self.maxConcurrency = maxConcurrency
    }

    public func runTask<T>(_ task: @escaping () async throws -> T) async throws -> T {
        if activeTasks >= maxConcurrency {
            await withCheckedContinuation { continuation in
                continuations.append(continuation)
            }
        }

        activeTasks += 1
        defer { activeTasks -= 1 }

        let result = try await task()

        if let continuation = continuations.first {
            continuations.removeFirst()
            continuation.resume()
        }

        return result
    }
}

public class SerialTransactionExecutor {
    private let cacher: CachingTransactionExecutor
    private let queue = SerialQueue()
    private var lastDigest: String?

    public init(cache: ObjectCache = ObjectCache()) {
        self.cacher = CachingTransactionExecutor(cache: cache)
    }

    public func execute(_ transaction: SuiTransaction) async throws -> TransactionResult {
        return try await queue.runTask {
            return try await self.cacher.execute(transaction)
        }
    }

    public func executeTransaction(_ transaction: SuiTransaction, signatures: [String]) async throws -> TransactionResult {
        return try await execute(transaction)
    }

    public func resetCache() {
        cacher.reset()
    }

    public func getCache() -> ObjectCache {
        return cacher.getCache()
    }
}

public struct CoinInfo {
    public let id: String
    public let version: String
    public let digest: String
    public let balance: UInt

    public init(id: String, version: String, digest: String, balance: UInt) {
        self.id = id
        self.version = version
        self.digest = digest
        self.balance = balance
    }
}

public class ParallelTransactionExecutor {
    private let maxPoolSize: Int
    private let executeQueue = ParallelQueue()
    private let cacher: CachingTransactionExecutor
    private var objectQueues: [String: [() -> Void]] = [:]
    private var coinPool: [CoinInfo] = []
    private var pendingTransactions = 0

    public init(maxPoolSize: Int = 50, cache: ObjectCache = ObjectCache()) {
        self.maxPoolSize = maxPoolSize
        self.cacher = CachingTransactionExecutor(cache: cache)
    }

    public func executeAll(_ transactions: [SuiTransaction]) async throws -> [TransactionResult] {
        return try await withThrowingTaskGroup(of: TransactionResult.self) { group in
            for transaction in transactions {
                group.addTask {
                    try await self.execute(transaction)
                }
            }

            var results: [TransactionResult] = []
            for try await result in group {
                results.append(result)
            }
            return results
        }
    }

    public func execute(_ transaction: SuiTransaction) async throws -> TransactionResult {
        let usedObjects = try await getUsedObjects(transaction)
        return try await executeQueue.runTask {
            let result = try await self.cacher.execute(transaction)
            self.pendingTransactions -= 1
            return result
        }
    }

    private func getUsedObjects(_ transaction: SuiTransaction) async throws -> Set<String> {
        return Set()
    }

    public func resetCache() {
        cacher.reset()
    }

    public func getCache() -> ObjectCache {
        return cacher.getCache()
    }
}

public struct TransactionExecutor {
    private var transactions: [SuiTransaction] = []

    public init() {}

    public mutating func execute(_ transaction: SuiTransaction) -> TransactionResult {
        transactions.append(transaction)
        return TransactionResult.success()
    }

    public func reset() {
        transactions.removeAll()
    }
}

File: java-sdks/src/main/java/com/suisdks/sui/transactions/executor/TransactionExecutor.java
package com.suisdks.sui.transactions.executor;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.util.concurrent.CompletableFuture;
import java.math.BigInteger;

class TransactionResult {
    private final boolean success;
    private final String error;

    public TransactionResult(boolean success, String error) {
        this.success = success;
        this.error = error;
    }

    public boolean isSuccess() {
        return success;
    }

    public String getError() {
        return error;
    }

    public static TransactionResult success() {
        return new TransactionResult(true, null);
    }

    public static TransactionResult failure(String error) {
        return new TransactionResult(false, error);
    }
}

class ObjectCache {
    private final Map<String, Map<String, Object>> cache = new HashMap<>();
    private final Map<String, Map<String, Object>> ownedObjects = new HashMap<>();
    private final Map<String, Map<String, Object>> customCache = new HashMap<>();

    public void set(String id, Map<String, Object> object) {
        cache.put(id, object);
    }

    public Map<String, Object> get(String id) {
        return cache.get(id);
    }

    public void delete(String id) {
        cache.remove(id);
        ownedObjects.remove(id);
        customCache.remove(id);
    }

    public void clearOwnedObjects() {
        ownedObjects.clear();
    }

    public void clearCustom() {
        customCache.clear();
    }

    public Map<String, Object> getOwnedObject(String id) {
        return ownedObjects.get(id);
    }

    public void setOwnedObject(String id, Map<String, Object> object) {
        ownedObjects.put(id, object);
    }

    public Map<String, Object> getCustom(String key) {
        return customCache.get(key);
    }

    public void setCustom(String key, Map<String, Object> value) {
        customCache.put(key, value);
    }

    public void deleteCustom(String key) {
        customCache.remove(key);
    }
}

class SerialTransactionExecutor {
    private final TransactionExecutor executor;
    private final Object lock = new Object();
    private final ObjectCache cache = new ObjectCache();

    public SerialTransactionExecutor() {
        this.executor = new TransactionExecutor();
    }

    public CompletableFuture<TransactionResult> execute(Transaction transaction) {
        return CompletableFuture.supplyAsync(() -> {
            synchronized (lock) {
                return executor.execute(transaction);
            }
        });
    }

    public void resetCache() {
        cache.clearOwnedObjects();
        cache.clearCustom();
    }

    public ObjectCache getCache() {
        return cache;
    }
}

class ParallelTransactionExecutor {
    private final int maxWorkers;
    private final ObjectCache cache = new ObjectCache();
    private final Map<String, List<Runnable>> objectQueues = new HashMap<>();

    public ParallelTransactionExecutor(int maxWorkers) {
        this.maxWorkers = maxWorkers;
    }

    public CompletableFuture<List<TransactionResult>> executeAll(List<Transaction> transactions) {
        return CompletableFuture.supplyAsync(() -> {
            List<TransactionResult> results = new ArrayList<>();
            for (Transaction transaction : transactions) {
                results.add(execute(transaction));
            }
            return results;
        });
    }

    private TransactionResult execute(Transaction transaction) {
        if (transaction == null) {
            return TransactionResult.failure("transaction is null");
        }
        if (transaction.getSender() == null || transaction.getSender().isBlank()) {
            return TransactionResult.failure("missing sender");
        }
        List<Map<String, Object>> commands = transaction.getCommands();
        if (commands == null || commands.isEmpty()) {
            return TransactionResult.failure("transaction has no commands");
        }
        String key = transaction.getSender();
        objectQueues.computeIfAbsent(key, ignored -> new ArrayList<>());
        if (objectQueues.get(key).size() >= maxWorkers) {
            return TransactionResult.failure("parallel queue saturated for sender");
        }
        Runnable marker = () -> {
        };
        objectQueues.get(key).add(marker);
        try {
            return TransactionResult.success();
        } finally {
            objectQueues.get(key).remove(marker);
        }
    }

    public void resetCache() {
        cache.clearOwnedObjects();
        cache.clearCustom();
    }

    public ObjectCache getCache() {
        return cache;
    }
}

class CachingTransactionExecutor {
    private final ObjectCache cache;

    public CachingTransactionExecutor(ObjectCache cache) {
        this.cache = cache == null ? new ObjectCache() : cache;
    }

    public CompletableFuture<TransactionResult> execute(Transaction transaction) {
        return CompletableFuture.supplyAsync(() -> {
            if (transaction == null) {
                return TransactionResult.failure("transaction is null");
            }
            if (transaction.getSender() == null || transaction.getSender().isBlank()) {
                return TransactionResult.failure("missing sender");
            }
            if (transaction.getCommands() == null || transaction.getCommands().isEmpty()) {
                return TransactionResult.failure("transaction has no commands");
            }
            cache.setCustom("lastSender", Map.of("sender", transaction.getSender()));
            cache.setCustom("lastCommandCount", Map.of("count", transaction.getCommands().size()));
            return TransactionResult.success();
        });
    }

    public void applyEffects(Map<String, Object> effects) {
        if (effects.containsKey("changedObjects")) {
            List<Map<String, Object>> changedObjects = (List<Map<String, Object>>) effects.get("changedObjects");
            for (Map<String, Object> changedObj : changedObjects) {
                String objectId = (String) changedObj.get("objectId");
                Map<String, Object> outputState = (Map<String, Object>) changedObj.get("outputState");
                if (outputState != null && outputState.containsKey("ObjectWrite")) {
                    cache.setOwnedObject(objectId, changedObj);
                }
            }
        }
    }

    public void reset() {
        cache.clearOwnedObjects();
        cache.clearCustom();
    }

    public ObjectCache getCache() {
        return cache;
    }
}

public class TransactionExecutor {
    private final List<Transaction> queue = new ArrayList<>();

    public TransactionExecutor() {
    }

    public TransactionResult execute(Transaction transaction) {
        if (transaction == null) {
            return TransactionResult.failure("transaction is null");
        }
        if (transaction.getSender() == null || transaction.getSender().isBlank()) {
            return TransactionResult.failure("missing sender");
        }
        if (transaction.getCommands() == null || transaction.getCommands().isEmpty()) {
            return TransactionResult.failure("transaction has no commands");
        }
        queue.add(transaction);
        return TransactionResult.success();
    }

    public void reset() {
        queue.clear();
    }
}

class Transaction {
    private List<Map<String, Object>> commands;
    private String sender;
    private BigInteger gasPrice;
    private BigInteger gasBudget;

    public Transaction() {
        this.commands = new ArrayList<>();
    }

    public List<Map<String, Object>> getCommands() {
        return commands;
    }

    public String getSender() {
        return sender;
    }

    public void setSender(String sender) {
        this.sender = sender;
    }

    public BigInteger getGasPrice() {
        return gasPrice;
    }

    public void setGasPrice(BigInteger gasPrice) {
        this.gasPrice = gasPrice;
    }

    public BigInteger getGasBudget() {
        return gasBudget;
    }

    public void setGasBudget(BigInteger gasBudget) {
        this.gasBudget = gasBudget;
    }
}
